src/CodeGen_Internal.cpp

```diff
@@ -407,6 +407,16 @@ Expr lower_euclidean_mod(Expr a, Expr b) {
     }
 }
 
+Expr lower_signed_shift_left(Expr a, Expr b) {
+  Expr b_unsigned = abs(b);
+  return select(b >= 0, a << b_unsigned, a >> b_unsigned);
```

Comment:
I'd modify these to do both simplification and then common_subexpression_elimination on the output. So something like:
```
Expr val = select(b >= 0, a << b_unsigned, a >> b_unsigned);
return simplify(common_subexpression_elimination(val));
```
Halide is better at proving things are positive than LLVM, so the switch might simplify away, and if a/b are large Exprs, duplicating them needlessly creates extra codegen work.

Reply:
(and assumes that llvm with successfully CSE them away again, which it almost certainly will, but still... in theory we have more information about what is and isn't pure than llvm)

Reply:
I added common_subexpression_elimination, but simplify seemed to segfault places. I'm not sure if that's expected? I can investigate more if that should have worked.

Reply:
That's not expected and is concerning. It would be good to investigate more.

Reply:
As a good rule of thumb, segfaults are never expected or acceptable -- we'll have to run this to ground and fix it.

Reply:
I think I figured out what was happening here. There was an infinite loop happening between Simplify_Call and the new stuff in LLVM_Internal, which I found a way around in Simplify_Call. I'm still trying to resolve some weird output related to Dillon's comment in IROperator, so I'll update the working code when that's resolved, too.

Reply:
Updated now with the fix. Please take another look when you get a chance!